core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -0,0 +1,136 @@
+   *
+   * @param table the table for which partition stats to be computed.
+   * @param snapshot the snapshot for which partition stats is computed.
+   * @return the collection of {@link PartitionStats}
+   */
+  public static Collection<PartitionStats> computeStats(Table table, Snapshot snapshot) {
+    Preconditions.checkArgument(table != null, "table cannot be null");
+    Preconditions.checkArgument(snapshot != null, "snapshot cannot be null");
+
+    StructType partitionType = Partitioning.partitionType(table);
+    if (partitionType.fields().isEmpty()) {
+      throw new UnsupportedOperationException(
+          "Computing partition stats for an unpartitioned table");
+    }
+
+    List<ManifestFile> manifests = snapshot.allManifests(table.io());
+
+    ExecutorService executorService = ThreadPools.getWorkerPool();
+    List<Future<StructLikeMap<PartitionStats>>> futures = Lists.newArrayList();
+    manifests.forEach(
+        manifest -> {
+          Future<StructLikeMap<PartitionStats>> future =
+              executorService.submit(() -> collectStats(table, manifest, partitionType));
+          futures.add(future);
+        });
+
+    StructLikeMap<PartitionStats> statsMap = StructLikeMap.create(partitionType);
+    for (Future<StructLikeMap<PartitionStats>> future : futures) {
+      try {
+        future
+            .get()
+            .forEach(
+                (key, value) ->
+                    statsMap.merge(
+                        key,
+                        value,
+                        (existingEntry, newEntry) -> {
+                          existingEntry.appendStats(newEntry);
+                          return existingEntry;
+                        }));
+      } catch (InterruptedException | ExecutionException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    return statsMap.values();
+  }
+
+  /**
+   * Sorts the {@link PartitionStats} based on the partition data.
+   *
+   * @param stats collection of {@link PartitionStats} which needs to be sorted.
+   * @param partitionType unified partition schema.
+   * @return the list of {@link PartitionStats}
+   */
+  public static List<PartitionStats> sortStats(
+      Collection<PartitionStats> stats, StructType partitionType) {
+    List<PartitionStats> entries = Lists.newArrayList(stats.iterator());
+    entries.sort(
+        Comparator.comparing(PartitionStats::partition, Comparators.forType(partitionType)));
+    return entries;
+  }
+
+  private static StructLikeMap<PartitionStats> collectStats(
+      Table table, ManifestFile manifest, StructType partitionType) {
+    try (ManifestReader<?> reader = openManifest(table, manifest)) {
+      StructLikeMap<PartitionStats> statsMap = StructLikeMap.create(partitionType);
+
+      for (ManifestEntry<?> entry : reader.entries()) {
+        ContentFile<?> file = entry.file();
+        PartitionSpec spec = table.specs().get(file.specId());
+        PartitionData key =
```

Comment:
I think I understand why it starts to fail if we use the existing `coercePartition` method. It fails because the entry object is reused, meaning `partition` we project is mutable.

Take a look [here](https://github.com/apache/iceberg/blob/72fd9ab9f44cfb3e4b08f849cf7b5132f84b9936/core/src/main/java/org/apache/iceberg/util/TableScanUtil.java#L156) at how we solve in `TableScanUtil`.

Reply:
> I think I understand why it starts to fail if we use the existing coercePartition method. It fails because the entry object is reused, meaning partition we project is mutable.

Thanks for this. 

> Take a look [here](https://github.com/apache/iceberg/blob/72fd9ab9f44cfb3e4b08f849cf7b5132f84b9936/core/src/main/java/org/apache/iceberg/util/TableScanUtil.java#L156) at how we solve in TableScanUtil.

They create two objects of `PartitionData` (one empty and one with copyFor) per `StructProjection`, Instead I used internal code to create one `PartitionData` object per `StructProjection`. So, I feel no need to modify this code. WDYT?